src/dos/cdrom_ioctl_linux.cpp

```diff
@@ -113,4 +320,159 @@ bool CDROM_Interface_Ioctl::ReadSectorsHost([[maybe_unused]] void *buffer,
 	return false; /*TODO*/
 }
 
+void CDROM_Interface_Ioctl::InitAudio(int device_number)
+{
+	if (mixer_channel) {
+		return;
+	}
+
+	std::string name = std::string(MixerChannelNamePrefix) +
+	                   std::to_string(device_number);
+	audio_buffer.resize(AudioBufferBytes / 2);
+
+	auto callback = std::bind(&CDROM_Interface_Ioctl::CDAudioCallback,
+	                          this,
+	                          std::placeholders::_1);
+	mixer_channel = MIXER_AddChannel(callback,
+	                                 REDBOOK_PCM_FRAMES_PER_SECOND,
+	                                 name.c_str(),
+	                                 {ChannelFeature::Stereo,
+	                                  ChannelFeature::DigitalAudio});
+}
+
+void CDROM_Interface_Ioctl::CDAudioCallback(uint16_t requested_frames)
+{
+	if (audio_position == 0) {
+		if (sectors_remaining < 1) {
+			StopAudio();
+			return;
+		}
+		cdrom_read_audio cd = {};
+		cd.addr.lba         = current_sector;
+		cd.addr_format      = CDROM_LBA;
+		cd.buf = reinterpret_cast<uint8_t*>(audio_buffer.data());
+
+		// Maximum this ioctl can read in a single call is 1 second (75
+		// redbook frames)
+		cd.nframes = std::min(sectors_remaining, CD_FRAMES);
+		constexpr int PcmFramesPerRedbookFrame = BYTES_PER_RAW_REDBOOK_FRAME /
+		                                         BYTES_PER_REDBOOK_PCM_FRAME;
+		frames_in_buffer = cd.nframes * PcmFramesPerRedbookFrame;
+		current_sector += cd.nframes;
+		sectors_remaining -= cd.nframes;
+		if (ioctl(cdrom_fd, CDROMREADAUDIO, &cd) != 0) {
+			// Write out silence if something goes wrong
+			std::fill(audio_buffer.begin(), audio_buffer.end(), 0);
+
+#ifdef DEBUG_IOCTL
+			LOG_WARNING("CDROM_IOCTL: CDAudioCallback: CDROMREADAUDIO ioctl failed");
+#endif
+		}
+	}
+
+	const size_t frames_available = frames_in_buffer - audio_position;
+	const size_t frames_to_write = std::min(static_cast<size_t>(requested_frames),
+	                                        frames_available);
+
+	// audio_position must be doubled since a frame is two int16_t's for stereo sound
+#ifdef WORDS_BIGENDIAN
+	mixer_channel->AddSamples_s16_nonnative(frames_to_write,
+	                                        audio_buffer.data() +
+	                                                (audio_position * 2));
+#else
+	mixer_channel->AddSamples_s16(frames_to_write,
+	                              audio_buffer.data() + (audio_position * 2));
```

Comment:
Yeah, so `audio_position * 2` would simply become `frame_pos` as your buffer would be a vector (or `std::array`, up to you) of `AudioFrame`s.

Reply:
I renamed `audio_position` to `frame_position` but it still has to be multiplied by 2 because the vector is in samples.  See my above comment about `AudioFrame`.  Need some clarification there.

Reply:
As per my long explanation. You'll have this instead:

```
std::vector<AudioFrame> audio_buffer = {};
```

Then you simply pass yout `int16_t` samples to this contructor:

```
constexpr AudioFrame(const int16_t l, const int16_t r)
```

And push the AudioFrames into the vector. That's it. Make sure to remove every trace of referring to "samples"; the code should only refer to "frames" after this.

Reply:
>  And push the AudioFrames into the vector. Add the results to the channel with AddSamples_sfloat. That's it.

`AddSamples_sfloat` won't take a vector of `AudioFrame` is the problem.  I'd be creating a vector that I can't really do anything with.  The add samples function just wants a raw pointer to some floats.

Reply:
Just do this like every other audio device that uses `AddSamples_sfloat`. Look up the usages, please. `audio_frames` is a `std::vector<AudioFrame>` in this example:

```cpp
channel->AddSamples_sfloat(requested_audio_frames, &audio_frames[0][0]);
```

Again, look at the bigger picture. When everything in the codebase will only push floats to the mixer via `AddSamples_sfloat` and `AddSamples_mfloat`, we can remove that huge templated `AddSamples` junk and introduce two much simpler `AddSamplesStereo` and `AddSamplesMono` functions or something that just take `std::vector<AudioFrame>` directly.

One step at a time...

Reply:
Okay, I'm kinda busy now, this is surprising. There shouldn't be _that_ much of a performance difference... but then, debug builds are naturally 4-5x slower, or a lot more in some cases. Consider switching to `std::array` perhaps? But probably you'd experience the same relative slowness in debug builds because of the bounds checks or something.

At the end of the day, only the performance of the release build matters.

Reply:
Btw, only use emplace, that's the potentially more performant version as it skips creating the temporary.

Don't use debug builds for judging the performance of any code, it's meaningless.

Reply:
So I was definitely measuring the wrong thing when I said "release builds are mostly the same".  That was wrong.  I did a micro-benchmark of just the loop that was converting 1 second of audio from int to float.

Debug with push_back - 3ms
Debug with emplace_back - 10ms
Release build with either (times are about the same) - 150 **microseconds**

So yeah... almost two orders of magnitude slower in the worst case.  I see your point about not trusting debug performance now.

I think this won't ever be a problem with release builds based on that but I will go ahead and take your suggestion to do the conversion as needed.  If anything, it saves a bit of memory.